### PR TITLE
[GISel][RISCV] Legalize G_CONSTANT_FOLD_BARRIER

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch supports `G_CONSTANT_FOLD_BARRIER` on RISCV to generate the following inst seq without crash:
```
define i64 @xor_and_i64(i64 %x) {
entry:
  %y = and i64 %x, 16383
  %z = xor i64 %y, 16368
  ret i64 %z
}
```

I don't know how to avoid creating `G_CONSTANT_FOLD_BARRIER` with s64 on rv32 :(

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=riscv32 -run-pass=legalizer %s -o - | FileCheck %s
+---
+name:            xor_and_i32
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir:4]: Testname is misleading, also you just need the one instance of the instruction

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir:4]: > you just need the one instance of the instruction

Generating `CONSTANT_FOLD_BARRIER` requires at least two constants.

Comment (arsenm) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir:4]: You've bypassed the creation condition and are writing the instruction directly though

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir:4]: > You've bypassed the creation condition and are writing the instruction directly though

I generated these tests with `bin/llc -mtriple=riscv64 -stop-before=legalizer -global-i
sel -o - test.ll`. Is it correct?

Comment (arsenm) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir:4]: You can just hand write MIR, especially if it's just one instruction

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -227,7 +227,7 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
   ConstantActions.widenScalarToNextPow2(0).clampScalar(0, s32, sXLen);
 
   // TODO: transform illegal vector types into legal vector type
-  getActionDefinitionsBuilder(G_IMPLICIT_DEF)
+  getActionDefinitionsBuilder({G_IMPLICIT_DEF, G_CONSTANT_FOLD_BARRIER})
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:230]: Looks like you are supporting vector types but there is no vector types tests.

**Review by michaelmaitland — COMMENTED** [review]
Do you plan on doing regbankselect and instructionselect in a different patch?

**Comment by michaelmaitland** [comment]
Do you know why the `G_CONSTANT_FOLD_BARRIER` was created in the first place?

**Comment by topperc** [comment]
> Do you know why the `G_CONSTANT_FOLD_BARRIER` was created in the first place?

I think it is created for constants hoisted by the ConstantHoisting pass.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-constbarrier-rv32.mir]: Fixed.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:230]: Done.

**Comment by dtcxzyw** [comment]
> Do you plan on doing regbankselect and instructionselect in a different patch?

No. X86 and AArch64 backends only mark G_CONSTANT_FOLD_BARRIER as legalized. I believe regbankselect and instructionselect for this node have been handled by the generic GISel logic.

**Review by arsenm — APPROVED** [review]

**Pull request merged**